Sensemaker store refactor round 2

pospi opened the pull request:
Builds off #33 to derive core `SensemakerStore` API via derived functional streams.

Also includes additional fixes to the test harness in order to facilitate unit testing of the client APIs.

pospi committed:
fix Holochain version pinning

pospi committed:
export interface types for referencing in application TS code

pospi committed:
update lockfiles

pospi committed:
import immutable & reactive data libraries

pospi committed:
expose API response type for referencingg in Store & external libs

pospi committed:
first pass rewrite of SensemakerStore internals to invert data model for reactive streams

Major changes:

- `_resourceAssessments` no longer tracks data for all resources as a singleton, and
  is now composed of separate observables for each `resource_eh` that are merged into
  a single (flat) stream at query time.
  **THIS IS A BREAKING CHANGE** from the older "nested object" model, but allows more
  deliberate bindings into the store data.
- `immer` now manages all updates to data, ensuring that objects are efficiently
  updated and immutably re-referenced upon changes.

pospi committed:
suggested method name change

pospi committed:
implement fine-grained SensemakerStore observable management & derivation helpers

pospi committed:
whitespace fixes

pospi committed:
fix call signatures on getAssessment & update to sync to cached Assessments upon fetching

pospi committed:
remove test runner workaround, fix TS config and reorganise test deps for direct compat with Tape

pospi committed:
define types to pass compiler checks

pospi committed:
add module clean commands

pospi committed:
update lockfile

pospi committed:
fix helper imports

pospi committed:
change test commands & config to use tsx/ESBuild packages for compilation with separate typechecking

works around bugs in TS compiler: @see https://github.com/microsoft/TypeScript/issues/41567

pospi committed:
update lockfile

weswalla committed:
Merge branch 'develop' into feature/sensemaker-composable-streams

weswalla commented on client/src/sensemakerService.ts:
```diff
@@ -26,14 +28,14 @@ export class SensemakerService {
   }
 
-  async getAssessment(assessmentEh: EntryHash): Promise<HolochainRecord> {
+  async getAssessment(assessmentEh: EntryHash): Promise<Assessment> {
```
should we revert this back to `HolochainRecord` until we change the return type of the `get_assessment` zome function to return an assessment?

adaburrows replied:
Probably. This would seem confusing if the code isn't actually returning this object.

adaburrows replied:
I'm actually confused. What is the difference between the HolochainRecord and the actual typed value? Does it matter in the client? In the playspace, I'm doing this same thing and it works just fine. On the other hand, in the zome code, I do have to do some HDK calls and pattern matching to get the correct Rust types.

weswalla replied:
the difference is that the `Record` contains the `Action` and the `Entry`, and the `Assessment` is serialized and thus needs to be decoded like: `decode((tenScaleRangeReadOutput.entry as any).Present.entry) as any`

adaburrows replied:
Do you mean `decode((tenScaleRangeReadOutput.entry as any).Present.entry) as Assessment`?

Is `tenScaleRangeReadOutput.entry?.Present?.entry` not already decoded?

Is `decode` actually necessary? It would be necessary if msg pack does not automatically recurse in its decoding, as `decode` is automatically called in the client through a very roundabout chain of calls: [`callZome`](https://github.com/holochain/holochain-client-js/blob/71c8c6292aa1fe98b7d79af99820757ad4c3ad10/src/api/app/websocket.ts#L129) -> [`_requester`](https://github.com/holochain/holochain-client-js/blob/71c8c6292aa1fe98b7d79af99820757ad4c3ad10/src/api/app/websocket.ts#L96) -> [`requestTransformer`](https://github.com/holochain/holochain-client-js/blob/71c8c6292aa1fe98b7d79af99820757ad4c3ad10/src/api/common.ts#L34) -> [`callTransformer.output`](https://github.com/holochain/holochain-client-js/blob/71c8c6292aa1fe98b7d79af99820757ad4c3ad10/src/api/app/websocket.ts#L185) -> [`decode`](https://github.com/msgpack/msgpack-javascript/blob/6e22414e88afb165b0e29c797272e6e9aa0655b1/src/decode.ts#L24) -> [`Decoder.decode`](https://github.com/msgpack/msgpack-javascript/blob/6e22414e88afb165b0e29c797272e6e9aa0655b1/src/Decoder.ts#L185)

adaburrows replied:
Oh, I just read the parentheses wrong.

weswalla commented on client/src/sensemakerStore.ts:
```diff
@@ -1,29 +1,69 @@
+
+// TypeScript interface
+
+type AssessmentObservable = Observable<Set<Assessment>> & Writable<Set<Assessment>>
```
I'm not super familiar with TS intersection types, could you elaborate on why we need to do this?

adaburrows replied:
This basically says that the objects we are passing around conform to both the RXjs Observable and the Svelte Store's Writable interface over an underlying type of `Set<Assessment>`.

I'd actually want to know why we are requiring both in all cases. Wouldn't we want
```
type ReadableAssessmentObservable = Observable<Set<Assessment>> & Readable<Set<Assessment>>
type WritableAssessmentObservable = Observable<Set<Assessment>> & Writable<Set<Assessment>>
```
depending on if we need an input stream or output stream?

pospi replied:
I think `Writable` is a superclass of `Readable`. But we should add some unit tests around readables specifically and ensure everything works as expected.

adaburrows replied:
It is a superclass, but my nit-picky brain thinks that if you properly want to reason about the functors between structural types, you need to know the exact minimum interface necessary for each use of it. So do we need to ensure that something passed in has the `Readable` (`subscribe`) interface, or do we need something that is also `Writable` (with `set` and `update`)?

pospi committed:
made resource & dimension parameters to load API optional

pospi committed:
target tests compilation to latest ESModule spec

pospi committed:
swap package mgr lockfiles

pospi committed:
update package manifests to switch to PNPM, fixing Typescript type incompatibility errors

pospi committed:
update flake lockfile

pospi committed:
add temporary adapter code to SensemakerService to return Assessments in flat array

pospi committed:
rewrite internal SensemakerStore architecture to use single stream of Assessments as core data store

pospi committed:
compile to more modern JS version for use of ES6 flatMap method

pospi committed:
add test mocking helpers for SensemakerStore client API unit tests

pospi committed:
first simple test for Assessment loading functionality

pospi committed:
update lockfile

pospi commented:
Really happy with the testing situation. This is what I'm running from the toplevel of the repository to check correctness as I go (one test file at a time). Executes almost instantly compared to the zome API tests (:

```sh
npm run build:client && pushd tests; node_modules/tape/bin/tape -r ./node_modules/tsx/dist/preflight.cjs src/sensemaker_client/assessment_loading.ts; popd
```

pospi committed:
implement some interfaces to update mocked API data during test process

pospi committed:
extend Assessment loading tests to verify progressive additions to SensemakerStore

pospi committed:
fix test assertions not comparing ES6 Set & Map in emitted values correctly

pospi committed:
comment out currently failing methods, pending reimplementation

pospi committed:
update lockfile

pospi committed:
change assessment test for additional readability of mocked data relationships

pospi committed:
switch to alternate implementation using stateful observable for complete set of loaded Assessments

pospi committed:
update test expectations to match revised internal logic for resourceAssessments()

pospi commented:
Another couple of updates to earlier that might help to explain my confusion @adaburrows.

[6d08f43](https://github.com/neighbour-hoods/sensemaker-lite/pull/41/commits/6d08f43405c3937c632150f83552f1c5f4bdaea2) modifies the internals again such that "resourceAssessments" is intended to persistently return the set of all `Assessments` ever loaded.

Instead in my according test expectations of [8dd0867](https://github.com/neighbour-hoods/sensemaker-lite/pull/41/commits/8dd0867763045b28681b2436e9b913f0dde01e5d) I have 1/3 tests passing, with the 2nd and 3rd tests only receiving the last emitted `Assessment` in the set. When I tap the streams I can see that it's working correctly *somewhere* in that byzantine conduit with a set of length 2 being emitted successfully after the second call to `store.loadAssessmentsForResources`. But then for whatever reason `_resourceAssessments` emits *again*, and some new version of `_allResourceAssessments` pops into being and emits a set of length 1.

Is this potentially related to how stream subscriptions are handled after a stream is deemed to have been drained? Are there ways that I can override any such behaviour so that subscribers will listen to streams indefinitely?

weswalla committed:
add specific test scripts for client or sensemaker

pospi committed:
move rx TestScheduler init into shared module and improve output of test failures

pospi committed:
Merge branch 'feature/sensemaker-composable-streams' of github.com:neighbour-hoods/sensemaker-lite into feature/sensemaker-composable-streams

weswalla commented:
> Is this potentially related to how stream subscriptions are handled after a stream is deemed to have been drained? Are there ways that I can override any such behaviour so that subscribers will listen to streams indefinitely?

@pospi sadly this is beyond my knowledge.

adaburrows commented:
> Is this potentially related to how stream subscriptions are handled after a stream is deemed to have been drained? Are there ways that I can override any such behaviour so that subscribers will listen to streams indefinitely?

I think I'm confused about what you want to accomplish here. You're taking a result of an array of `Assesment`s and piping each `Assesment` individually. Then you're reconstituting a new `Set<Assesment>` each time you process one `Assement` (and I don't think Immer's `produce` helps anything in this case, since it's expecting an `Object` hash map to make changes over a persistent data structure). This seems excessive. It would be best to use an Object hash to index items by hash id when you use `produce`.

Do you want a stream processing system that only returns the next items queued which the subscriber hasn't read yet? Or do you want an array/index of all the records? Because right now, having `_allResourceAssessments` as an `Observable` means that when a subscriber reads the last emitted value, it is now empty for that value. Since I haven't read through the RxJS code, I don't actually know what the exact behavior is for this kind of scenario, but I would assume that each change to `_allResourceAssessments` would actually show up as separate events that will be pulled off the queue by each subscriber. As each object is pulled off, by the subscriber, it gets the next most recent value (the persistent data structure of merged records) piped through the `Observable`. Each new subscriber gets the full stream of updates from the beginning.

I think what you actually want for `_allResourceAssessments` is a [`Subject`](https://rxjs.dev/guide/subject) and not an `Observable`. You probably want that for `_resourceAssesments`, too. You probably also want to process each array of `Assessments` as one event. On the other hand, you could get rid of one of those intermediaries and just pipe in merged records in the one remaining `Subject`.

Or, you could dispense with RxJS completely and just use the Svelte `Stores`'s `update` method to cause it to emit an event to its subscriber list each time it updates. Then anything downstream can just use `derive` to create new forms of the data or `subscribe` to get the updates for when something has changed. This would honestly be the easiest way and wouldn't require people to be familiar with the intricacies of RxJS.

pospi commented:
> you could dispense with RxJS completely

> This would honestly be the easiest way and wouldn't require people to be familiar with the intricacies of RxJS.

Well, yes and no. I agree with not requiring people to be familiar with RxJS as a goal of the `SensemakerStore` API. The intention is use of `svelte-fuse-rx` to get the best of both worlds, since `rxWritable` inherits from `BehaviourSubject` (which [is a `Subject`](https://github.com/ReactiveX/rxjs/blob/6e3e5e49cb735289a8e8796dc58e682468a36028/src/internal/BehaviorSubject.ts#L9)) whilst maintaining compatibility with Svelte's store API.

So `_allResourceAssessments` and `_resourceAssesments` are both `Subjects`, tick.

> anything downstream can just use `derive` to create new forms of the data or `subscribe` to get the updates

This is the approach we've taken to date. Unfortunately what it means is that we kick the can further down the road and access to slices of `Assessment` data with common utility becomes a complex set of manual category-theoretic operations that in practise have been inefficient and error-prone when attempted with Svelte's APIs and the current storage format.

Ergo, introducing Rx operators.

> I'm confused about what you want to accomplish here.

Fair enough 'cos so am I really?   But I think what we want is something that wraps up complex Rx operations into convenience helper methods which give easy access to slices of the Sensemaker data so that nobody from the outside needs to consider the complexity of them and can just deal with & bind to them as regular Svelte data sources. (Bonus in this case of course is that they can be further manipulated with Rx piping.) I figured the easiest way of doing that was to have a single stream as the original source of truth: hence, `_resourceAssessments` and

> taking a result of an array of `Assesment`s and piping each `Assesment` individually.

Check. The intention was that this form the single source of truth for all loaded `Assessments`. I think the stream should emit all of them individually in the order they're loaded?

Additionally I believe we should be able to implement `assessmentsFilterOpts`, `assessmentsForResource`, `assessmentsForResourceDimension` and `latestAssessmentOf` as derivations of this original stream; as well as currently unimplemented helpers for `assessmentsForDimension`, `assessmentsForResourceDimensions` (plural) & `latestAssessmentForResourceDimensions` - all of which would be useful to different Sensemaker and Provider UI controls.

`_allResourceAssessments` is supposed to take a `scan` across the original stream and accumulate all of the loaded `Assessments` into a persistently cached `Set`. It seems to *sortof* work, but can't get it to pass tests.

(The `immer` stuff is probably an over-optimisation but it's nice in that `produce` will maintain strict equality if adding an `Assessment` that is already loaded. I believe it also fires only 1 event after a stream has settled, meaning that repeated calls to `next` (Rx) or `set` (Svelte) should have all queued updates emitted as a single UI redraw operation.)

> do you want an array/index of all the records?

Yes. I want `_allResourceAssessments` to continually emit an up-to-date version of that `Set`.

> right now, having `_allResourceAssessments` as an `Observable` means that when a subscriber reads the last emitted value, it is now empty for that value

This might be the thing I am missing about streams.

Also wondering if this still applies given that `_allResourceAssessments` is actually a `BehaviourSubject`?

> I would assume that each change to `_allResourceAssessments` would actually show up as separate events that will be pulled off the queue by each subscriber

That *sounds* like it might be what I'm seeing. Please say more?

pospi commented:
Hm I think the issue might actually be the choice of wrapper in `svelte-fuse-rx`. Of the [types of available `Subject`](https://www.learnrxjs.io/learn-rxjs/subjects), I probably want a `ReplaySubject` rather than a `BehaviourSubject`.

pospi committed:
update test to load multiple Assessments initially (which also seems to be failing with BehaviorSubject)

pospi committed:
test should assert two values published through stream during init, since Assessments notify separately

pospi committed:
rewrite SensemakerStore internals again using experimental ReplaySubject derivative

pospi committed:
expect all prior states in second stream loading assertion

pospi committed:
cache last output of resourceAssessments when returning to new subscribers to avoid unnecessary source stream recomputation

pospi committed:
should only emit latest values when asked

pospi committed:
update test description

pospi committed:
cache Set of all loaded Assessments so that subscribers are not required to compute full state from source _resourceAssessments observable

pospi committed:
update lockfile

pospi commented:
Well, [I implemented `ReplaySubject`](https://github.com/pospi/svelte-fuse-rx/blob/c8a5d72bf2133c8c4f1eeb8bd672f4acca4a6be3/src/lib/rx.ts#L36). What I have now are two possible and equally awkward situations.

See [tests for context](https://github.com/neighbour-hoods/sensemaker-lite/blob/377a09b0f8951e0e32abdd78471f711e2e375df2/tests/src/sensemaker_client/assessment_loading.ts#L21), specifically the 3 assertions on lines 34, 46, 57.

### With a `BehaviorSubject`

Multiple repeated calls to `.next()` from within `syncNewAssessments` only emits the last `Assessment` passed. The first test (L34) is thus unable to pass; so unless this is an incorrect method for initially populating a stream then `BehaviourSubject` appears to be a dead-end.

### With a `ReplaySubject`

`reduce`, `takeLast` and other operators which wait for the input stream to close are rendered useless, since the stream never closes. When these are used in the pipeline we get no output in tests.

@adaburrows does this match your understanding RE `ReplaySubject`'s behaviour?

`scan` and other operators which continually emit in response to input show data accumulating as expected when `tap`ped. First assertion (L34) is still failing but this appears to be a consequence of the test runner strictness:

    expected: |-
      [ { frame: 0, notification: { kind: 'N', value: Set (1) {{..}}, error: undefined } }, { frame: 1, notification: { kind: 'N', value: Set (2) {{..},{..}}, error: undefined } } ]
    actual: |-
      [ { frame: 0, notification: { kind: 'N', value: Set (1) {{..}}, error: undefined } }, { frame: 0, notification: { kind: 'N', value: Set (2) {{..},{..}}, error: undefined } } ]

To me the `frame: 0`|`1` difference indicates a successful 'append and update previously emitted stream value' behaviour. This appears to be a positive step forwards. Unsure if/how to test for this with Rx's `subscriptionMarbles` (I could simply change the assertions to ignore `frame` or reduce the stream frames prior to [comparison](https://github.com/neighbour-hoods/sensemaker-lite/blob/adf7c83205c04c226fac386b51723a12977644a5/tests/src/store_mocks.ts#L15) :thinking:)

@adaburrows does this seem reasonable given Rx's expectations regarding the behaviour of `ReplaySubject`? This is just a test runner rawness issue, right?

But before we get too carried away, the second assertion passing (L46) is more interesting:

- `store.loadAssessmentsForResources` on L40 shows the expected propagation of `a3` through `observed` in tapping. However the next `testScheduler.run` was replaying the `observed` stream from the start, emitting all old aggregate values before the final one. Inefficient.
- Adding `shareReplay(1)` to the pipeline seems to result in only pre-computed aggregate values being updated. Now when `testScheduler` runs against the original `observed` on L47, the output stream instantiated by `store.resourceAssessments()` is the only stream tapped. The source `ReplaySubject` is no longer unnecessarily being rewound. Fixed?

When the original stream is cached as an instance member of `SensemakerStore`, this caching via `shareReplay(1)` also appears to service as an efficient means for UI components to subscribe to "the current `Assessments` which have been loaded"; vs "processing all of the state transitions which have occurred to determine the currently loaded `Assessments`".

Thus, L57 now also passes. 2/3...

pospi commented:
Also to note that the ideal situation would be for L34 to emit a single value of `new Set([a1, a2])` (rather than `new Set([a1])` initially and then correcting itself); but I'm not sure whether this is important to "normal" subscribers such as UI components or just a `TestScheduler` quirk.

pospi committed:
expose method for unmounting SensemakerStore to close all open stream subscribers

pospi committed:
improve Rx test runner: compare against latest stream frames only, better comparison output

pospi committed:
update test expectation per new assertion method

pospi committed:
fix hash mocking method

pospi committed:
make mockAssessment filler args more terse

pospi committed:
reimplement Assessment filtering API

pospi committed:
add tests for Assessment filtering

adaburrows commented:
If you want to make it work the way I think you want it to, you likely need a [higher-order `Observable`](https://rangle.io/blog/thinking-in-nested-streams-with-rxjs) that listens for `Observable` streams and only emits a value once the input `Observable` is completed. Each time a new set of input `Assessment` values come in from some source, it will create a new `Observable` that completes as an input to the higher-order `Observable`.

Currently you have one stream that inputs into another stream. You are processing the single value emitted by the stream and immediately emitting another value on another stream that is an incomplete computation. You either have to make the second stream a higher-order stream OR you have to just process all input arrays before emitting an event. RxJS does not magically change the dynamics of the event emitters and the observer pattern, and this is why I don't see it as necessary.

Even if you switch to higher-order `Observable`s, there's still some wasted computation happening with the filtering you are doing to match on `EntryHashB64`s that could be saved by retaining the index that already exists in the result from the zome. Yes, you could argue that the bloat of keeping the hashes around as an index is not worth it, but you need to implement better identity maps than you currently have, memory is cheap, and it can be optimized for large numbers of `Assessments` later. How many values will be stored in memory for the foreseeable future? If you really want to search over the values with a `find`, the identity arrows I have below can be used for that, too.

Since you are using a `Set` you are inadvertently adding objects twice if `getAssessment` is called multiple times since you have no map between incoming `Assessment` objects and the ones that also already exist in the set. You also aren't checking to see if there's already a cached version of the `Assessment` in the store before adding the result. The `Set` uses object identity and you don't have a guarantee of object equivalence between updated versions of an object since each API call returns a new object since `({} === {}) === false`.

What I would do to make things happen quickly is place incoming objects in a hash table of hash tables so that the objects can be efficiently mutated and queried by the `EntryHashB64` of the `HolochainRecord` or `Assessment` in the future. The type signatures and some of the needed functions look like this:

```ts
export type ResourceAssessmentsResponse = Record<EntryHashB64, Assessment[]>
type AssessmentMap = Record<EntryHashB64, Assessment>
type RecordAssessmentMap = Record<EntryHashB64, AssessmentMap>

// XXX: Can we do this without modifying the return type from the zome?
type AssessmentIDArrow = (assessment: Assessment) => EntryHashB64

const RecordIDArrow = (record: HolochainRecord): EntryHashB64 => {
  return encodeHashToBase64(record.signed_action.hashed.hash);
}

const AssessmentRecordIDArrow = (assessment: Assessment): EntryHashB64 => {
  return encodeHashToBase64(assessment.resource_eh);
}

// Functions that satisfy this type signature should update the inner hash table of `Assessments` indexed by `EntryHashB64`
type AugmentAssessmentMap = (
  assessment: Assessment,
  assessmentId: AssessmentIDArrow,
  assessmentMap: AssessmentMap
) => AssessmentMap

// Functions that satisfy this type signature should update the outer hash table indexed by `HolochainRecord` `EntryHashB64` and the inner hash table of indexed Assessments
// Technically, the parameter containing assessments could be relaxed to an array.
type AugementRecordAssessmentMap = (
  assessmentsByRecord: ResourceAssessmentsResponse,
  recordId: typeof RecordIDArrow,
  recordMap: RecordAssessmentMap,
  assessmentRecordID: typeof AssessmentRecordIDArrow,
  assessmentId: AssessmentIDArrow,
  augmentAssessmentMap: AugmentAssessmentMap
) => RecordAssessmentMap
```

Granted, you will probably implement a curried version of these functions that conform to this type signature. For a similar example of this, see the [Valueflows accounting module](https://github.com/lightningrodlabs/rea-playspace/blob/main/modules/valueflows-models/src/Accounting.ts#L203) (yeah, there's a few bugs with that due to some missing validation logic, but this is a simpler application of the same logic).

As you do a comprehension over the the array of `Assessments` you can update the objects efficiently by index while using the function that conforms to the `AugementRecordAssessmentMap` type. If you still need the objects to be immutable, you can use Immer's `produce` function in the functions that satisfy the `AugmentAssessmentMap` and `AugementRecordAssessmentMap` types.

After you update the the objects, your store can emit an event. Otherwise, you'll just be emitting the incomplete computation at each step if you do it inside of the comprehension inside of the function conforming to `AugementRecordAssessmentMap`. Of course, if you are using the higher-order `Observable`s with `svelte-fuse-rx` then it'll do it as soon as the lower-order input observable is completed.

Let me know if you want a walk through about this.

pospi committed:
draft reimpl resource dimension filtering helpers

pospi committed:
more abstraction of SensemakerStore helpers, start on grouped latestAssessmentsOfDimensions method

pospi committed:
allow override of Assessment timestmaps when mocking

pospi committed:
sensemaker convenience filtering API tests

pospi committed:
more Rx imports, abstract types & define for Dimensions Record observable

pospi committed:
implement latestAssessmentsOfDimensions to get nested Record structure of data for dims per record

pospi committed:
sort with fewer ops

pospi committed:
update Assessment loading tests to test indexed dimension loading API

pospi committed:
debug output fixes

pospi committed:
progressively simulate timestamps 1s into the future to give comparison ops something to do

pospi committed:
complete SensemakerStore stream accessors, with internal refactor to clean up helper methods for succinctness

pospi committed:
method docs

pospi committed:
abstract more stream grouping helpers

pospi committed:
add Assessment loading tests for Sensemaker dashboard APIs

pospi committed:
remove unneeded utils exports

pospi committed:
rewrite abstractions surrounding SensemakerService to string-encode all EntryHashes. Type updates to enforce distinction between Sensemaker hash types.

pospi committed:
update tests & runner to use string-based API for Assessment data

pospi committed:
remove unused imports in test

pospi commented:
Lots to cover in here. There might be some things to separate out and some assumptions being made about the maturity of the code to date, so let's be clear about distinguishing what are architecturally stream-related topics vs what are other concerns.

### Hash indexing

> filtering you are doing to match on `EntryHashB64`s that could be saved by retaining the index that already exists in the result from the zome

I decided it best that we make the `SensemakerStore` decode hashes to strings upon loading from the zomes and just not have any `Uint8Array` data exposed outside of `SensemakerService` at all. Implemented in the most recent commits. Strings are just way easier to compare and index by, and they support quick strict equality checks.

### Identity arrows

Appreciate these suggestions and in the most recent commits you'll notice the `SensemakerStore` API is starting to look as functional and terse as I'd hoped it could :slightly_smiling_face:

pospi commented:
### Strict equality checks & aggregate API return structure

> Since you are using a `Set` you are inadvertently adding objects twice if `getAssessment` is called multiple times. You also aren't checking to see if there's already a cached version of the `Assessment` in the store before adding the result.

Yeah, I just hadn't gotten to that optimisation yet and since it's largely unrelated to the core Rx data pipeline I'd been deferring it. For completeness-

- [x] Decide whether or not to group the output of `assessmentsForResource`, `assessmentsForResourceDimensions`, `latestAssessmentsOfDimensions`, `assessmentsForDimension`, `assessmentsForResourcesInDimension` and `latestAssessmentsForResourcesInDimension` by sub-types or return as flat arrays/Sets.

(Basically where I am currently using `asSet()` we could be leveraging `mergeGroup<EntryHashB64, Assessment>(newestGroupedAssessments, {})()` in other ways via a different callback replacing `newestGroupedAssessments`.)

Any thoughts appreciated as to what the most commonly expected interfaces might be.

pospi commented:
All these other topics aside we can now proceed to!:

### Client library core data architecture

It's been interesting reading your reflections on appropriate use of Rx. In the way that most proponents talk about it, it's usually considered as a single source of truth data store at the heart of an application. My read of typical reasoning is that if you are using Rx then you shouldn't be using Redux or Svelte stores or any other type of data persistency layer because you'll create multiple sources of truth and cache invalidation problems. Like how the versions of React Router that are really bad are the ones which integrate with data storage layers 'cos then the URL and the store can become conflicting sources of truth.

It's also true that some people think it's cool and normal to have languages where functions are the only datatype and numbers are represented by repeated recursions into a unit function; so I take such perspectives on Rx with a grain of salt as with anything.


This stated, the dynamics of the pattern (borrowing your terminology for something that sounds like what this is) now "feels right" to me, even if there are probably hidden gremlins:

- All operations start with some coarse linear filtering of the input source in `O:n(1)` complexity.
- Accumulating operations like `asSet` or `mergeGroup` (see store helpers) always happen at the end of a data pipeline, which should mean that they only update if their pre-filtered inputs change.
- Grouping and other stream splitting happens in the middle of the pipeline, and-
- `mostRecentAssessment` & others which perform comparisons & selective filtering only operate on subsets of relevant / grouped data, prior to any final accumulation.

So, at least in theory, this seems like it's achieving what I set out to do even if some of the internals might be unnecessarily computing and discarding updates in response to the methods I'm using to cache and propagate existing data.

> place incoming objects in a hash table of hash tables

I figured the best thing I could do from here is to write tests asserting the behaviour I expect, and we can go back and adjust internals (and potentially expectations) later if necessary. I've implemented methods useful in a Dashboard-type situation and some corresponding tests- I had wondered whether some of your advice on approach may be due to not yet seeing the need for dimensional slicing to be done both 'horizontally' and 'vertically'.

Either way, they're passing so far.

> After you update the the objects, your store can emit an event.

My concern here is that Shadow DOM (in Lit, Svelte etc) doesn't give us the fast redraw execution engine of a Virtual DOM (like in React). So I don't want a massive table with hundreds of rows to redraw *every row* just because a single row had its like button clicked- I actually want only the particular widgets within the row that care about "like" data to redraw. There needs to be something precise like `latestAssessmentOf` available to support this- effectively using Rx in place of what React's diffing VDOM used to optimise away for us so that only a single WebComponent's Shadow DOM be redrawn. I believe this should also assist us in writing more neatly encapsulated UI components with fewer dependencies that are thus easily configurable into L3 apps.

I want to add some additional tests around rendering performance before we do any other rewrites because that will help me to test whether my assumptions are correct; and in a way that is materially useful in making some performance guarantees to Applet developers.

> **Future challenge: reimplement `SensemakerStore` internals and pass all tests using `BehaviorSubject` rather than `ReplaySubject`.**



### Further optimisations & improvements

Just extracting some of the key concerns / suggestions from your above reply for reflection upon in future iterations.

> you likely need a [higher-order `Observable`](https://rangle.io/blog/thinking-in-nested-streams-with-rxjs) that listens for `Observable` streams and only emits a value once the input `Observable` is completed. Each time a new set of input `Assessment` values come in from some source, it will create a new `Observable` that completes as an input to the higher-order `Observable`.

> you need to implement better identity maps than you currently have

> How many values will be stored in memory for the foreseeable future?

I'd say for something that is usable in communities in the low-hundreds of humans we should be targeting in the order of 10-100K `Assessment` records as a starting point.

adaburrows commented:
> I decided it best that we make the SensemakerStore decode hashes to strings upon loading from the zomes and just not have any Uint8Array data exposed outside of SensemakerService at all.

So, this is an interesting thing, because if you think about what makes the quickest comparison in terms of filtering, it would be taking the `UIntArray8`s coming in and making a map using the object reference as the the key. Then as each assessment comes in, you can compare the bytes and if they are identical, you can replace the `UIntArray8` with the reference to the one used as key. Then at most, there is only one `UIntArary8` in memory, but many objects reference it in a form of a persistent data structure.

> Appreciate these suggestions and in the most recent commits you'll notice the SensemakerStore API is starting to look as functional and terse as I'd hoped it could 🙂

Cool. I'm glad you used those pointers.

> (Basically where I am currently using asSet() we could be leveraging mergeGroup<EntryHashB64, Assessment>(newestGroupedAssessments, {})() in other ways via a different callback replacing newestGroupedAssessments.)

Yes, that would be a little bit better. But there's still extra cruft from my perspective.

> Any thoughts appreciated as to what the most commonly expected interfaces might be.

I think that this isn't as flexible as we would want for the Dashboard. We would want to get a data structure of Assessments grouped by resource hash. Preferably, also having the computed assessments in a separate key so that showing it in the table is as simple as iterating over the the resources, merging in the data from the applet's API, and showing the computed assessments. The full applet API is yet to be determined (but we had an initial exploratory meeting about it yesterday), but we need to fetch the data and possibly a view from the applet to display it in the dashboard.

Unless I'm misunderstanding this, we'll still need to iterate over the resultant set to group by resource, or we'd have to create as many observables as there are resources (and each observable will do the computation over the whole set of data each time we set one up, meaning there's a bunch of wasteful repeated filtering going on).

> My read of typical reasoning is that if you are using Rx then you shouldn't be using Redux or Svelte stores or any other type of data persistency layer because you'll create multiple sources of truth and cache invalidation problems.

Sure, but much of the existing application is written using Svelte stores lit-svelte-stores and the interface you're exposing is an object that exposes the nearly the same subscriber interface, but which differs by the number of parameters passed into the subscribe function. See [Svelte `Readable<T>`](https://github.com/sveltejs/svelte/blob/master/src/runtime/store/index.ts#L26) versus [RxJS `Subscribable<T>`](https://github.com/ReactiveX/rxjs/blob/master/src/internal/types.ts#L85). The interfaces underlying these are nearly identical, except for the `Readable<T>` allowing a cache invalidation function to be passed in along with a callback.

Additionally, you previously said you didn't want anyone to have to know about the internals of RxJS implementation in this library, but you are also not implementing the correct `Readable<T>` interface that Svelte components might require, since there's an additional parameter that should be allowed so the subscribing component can know when to invalidate any caching it might have done. Though we don't use it in [lit-svelte-stores](https://github.com/guillemcordoba/lit-svelte-stores/blob/main/src/store-subscriber.ts#L34), because the dynamics of [`derived`](https://github.com/sveltejs/svelte/blob/master/src/runtime/store/index.ts#L165) are nearly identical to merging multiple RxJS observables.

- [ ] We should at least make a note of this somewhere so other devs can be aware of this if they see people mixing and matching Svelte stores and RxJS streams and expect the same behavior as a Svelte store.

> All operations start with some coarse linear filtering of the input source in O:n(1) complexity.

Do you you want O(1) or O(n) complexity? Your notation should only include the fastest growing term sans coefficient of the full behavior of the number of operations done by an algorithm, typically notated as T(n). It has no coefficients because it's whichever coefficient provides the best fit upper bound of a polynomial T that represents the actual number of operations as the size of the input tends toward infinity. O(n) would roughly signify the upper bound to a function f with the highest degree of 1. Θ(n) is a theoretical estimate of performance (there will be some pair of coefficients that determine the upper and lower bounds of a polynomial T of degree log_n(n). You can check a [list of definitions here](https://en.wikipedia.org/wiki/Big_O_notation#Family_of_Bachmann%E2%80%93Landau_notations).

The problem with your implementation is that each time someone requests and observable, it replays the whole stream and does the computation instead of processing the stream once and caching the current state. This happens because you're using the [ReplaySubject](https://github.com/pospi/svelte-fuse-rx/blob/feature/replay-subject/src/lib/rx.ts#L38) which replays all of the data points. This means it also keeps a cache of every assessment ever loaded, even if it was loaded more than once.

> the best thing I could do from here is to write tests asserting the behaviour I expect

This is always good practice. But it's also good to fully understand the code you're using and why it produces the output it does.

> So, at least in theory, this seems like it's achieving what I set out to do even if some of the internals might be unnecessarily computing and discarding updates in response to the methods I'm using to cache and propagate existing data.

No, not in theory, only in practice. I want to know, honestly, do you fully understand the data structures being created and run in your browser when you do this? How many of the other people who have to maintain this code will fully understand it?

> I had wondered whether some of your advice on approach may be due to not yet seeing the need for dimensional slicing to be done both 'horizontally' and 'vertically'.

I was just making the assumption that people would always want assessments grouped by resource. since that's the logical unit of everything that we see in the UX. **Every other bit of slicing and dicing could still be done on the pre grouped data structure.** I suppose we could have situations where someone wants to see the most interacted with resource across all dimensions, but even then, those assessments are all grouped by resource, so a simple reduce and filter operation is all thats needed.

Maybe, just maybe, there will be a situation where we want to show things by dimension, but given the discussions we've been having with UX, I don't see us doing anything but querying the dimensions and displaying information about them and maybe checking to see if they are being used to assess anything. This data can also just be generated as a reduce operation over the data.

Additionally, as long as we're processing sets of new data, then multiple indices can be built up to ensure fast O(1) to O(log n) look up time for the data. The data doesn't even need to be duplicated, because JavaScript uses object references so each index would point to the same objects.

Then all of the computation can happen once, and be cached until we receive new updates over signals.

> My concern here is that Shadow DOM (in Lit, Svelte etc) doesn't give us the fast redraw execution engine of a Virtual DOM (like in React).

Well, there's a bit to unpack within that statement. In React, there's a bit of a mechanism going on to tract the changes in a [persistent data structure](https://en.wikipedia.org/wiki/Persistent_data_structure) in memory. If you're Rich Harris, you [argue that this is pure overhead](https://svelte.dev/blog/virtual-dom-is-pure-overhead). I happen to agree with him. Most of the time, web developers are so ignorant of the mechanics of what is happening in the browser that using a framework like React, OM, [Fulcro](https://book.fulcrologic.com/#_immutable_data_structures), or other library that handles this stuff works out faster than what the dev would have naively done. React, OM, and Fulcro rely on persistent data structures for efficiently diffing the data. If any of the parts of the structure change, the root of the data structure is created anew and all of the old data are shallow copied and new objects are created for what has changed. This allow for efficient diffing by comparing object identity instead of recursing down through each leaf value in each node of the data structure. Then once the diff is efficiently computed one only needs to do the same operations one would always have to do with the slow DOM, but now you're only replacing what you need to replace instead of the naive way of redrawing a lot of DOM.

Lit actually assumes that people want to manage how re-rendering is done themselves. That's why **I was appalled** when I saw what lit-svelte-stores was doing and how it didn't provide much in the way of managing complexity. Then I've been working on other projects that have taken my attention away from this project. But I've come to see that this isn't a bad thing, because it just means you have full control over when the `render()` method is called on any component. One just has to be judicious about how rendering is done.

So I'll be adding a bit of an example of how to manage this for something as complicated as the table I'm building. I [have to do this ticket I put in my ticket backlog](https://github.com/adaburrows/table-web-component/issues/4) for the table, because even with all of the "functional" code you wrote here, the whole table will re-render each time something is changed because I didn't want to think through the whole problem I found back when I wrote that table component.

In the case of Svelte, all the code is transformed into code that basically does only the final result of changing only what needs to be changed in the DOM and nothing more. It's actually pretty nifty. I don't know if it needs to interface with data that is nicely delineated and easy to understand when it changes. I would hope that at that interface it either checks for object identity (through requiring persistent or immutable data structures) or by comparing cached values with the presumed new value. I honestly need to learn more about how the generated code would interact with any of our data stores.

> I want to add some additional tests around rendering performance before we do any other rewrites because that will help me to test whether my assumptions are correct; and in a way that is materially useful in making some performance guarantees to Applet developers.

Unfortunately, you can't assume that developers will take advantage of any of the features you provide to attempt to make things a little more render efficient, especially with the issues around the lit-svelte-store decorators. However you can encourage devs to slice data into the smallest amount necessary to update the smallest regions of the screen.

Also, as stated above, this wont help anything around table performance, because I need to redesign the interface around that to return row stores (a simple projective operator on the full data set used to created a `derived` store that only fires an even if it has changed compared to the previous value when an event from the parent store fires) and cell stores that are then used only run the synthetic field derivations if there was actually a change.

pospi committed:
Merge remote-tracking branch 'origin/develop' into feature/sensemaker-composable-streams (partial, test typing work needs followthrough)

pospi committed:
typing & serialisation pass at core API bindings, all hashes are now managed in base64

pospi committed:
update lockfiles

pospi committed:
more typing fixes for tests

pospi committed:
made client optional for constructing SensemakerStore and overriding service, mostly for test mocking

pospi committed:
fix AppletConfig method loading to not mutate existing objects, rename 'check' method to 'load'

pospi committed:
update method name in docs

pospi committed:
fix AppletConfig deserializing

pospi committed:
implement missing createRange method on SensemakerService

pospi committed:
connect AppAgentInterface in test harness spinup for use in SensemakerService connection

pospi committed:
update applet_config tests to test SensemakerService adapter logic as well as zome calls

pospi commented:
### strings and Uint8Arrays

> if you think about what makes the quickest comparison in terms of filtering

I do, and it's a weird one to think about because [YavaScript](https://www.destroyallsoftware.com/talks/the-birth-and-death-of-javascript) is weird. Since `Uint8Array` is treated as a contiguous block of raw bytes starting at some memory address in an instance-like fashion similar to the ES5 `Array`; equality comparators don't work as expected and `([] === []) === false`.

On the other hand, strings are treated as primitives and therefore treated as pointers to heap-allocated data. `("" === "") === true` occurs because these two empty strings are actually *the same memory address* (`""` occurs exactly *once* in any ECMAScript interpreter); and JavaScript implements the `===` operator cheaply as a simple comparison of memory addresses.

In any case bbdcd62 and some followups have now moved even the core `SensemakerService` API to a string-based format which simplifies and tightens a lot of method signatures. It's a little verbose but I think I like it?

pospi commented:
Will respond to more of this later, it's a lot to take in.

For now I'm back to all tests passing; updated some in 82cc614 to test at `SensemakerService` level instead of raw zome calls. But I think it's still nice to have both.

Have also made updates to `AppletConfig` handling in 21123cf to rename methods & correctly propagate updates through to subscribers.

Next steps I think are:

- [ ] update to @nick-stebbings' Vite-based build pipeline in order to get component tests operating; then I can finish this PR off with tests for rendering efficiency.
- [ ] adjust return values of methods returning flat `Assessment` arrays to return into sub-slices grouped by Record and/or Dimension.

adaburrows commented:
@pospi If your step about `adjust return values of methods returning flat Assessment...` is about the back end at all, I think we should wait to adjust the code until we merge in my really big changes to the zome. Also, it will change some of what you need to do here.

There's also some caching strategies and summarization strategies we should talk about since I think we may be requesting and caching more data from the Sensemaker than strictly necessary.

pospi committed:
WIP optimisation of SensemakerStore reactive data pipeline to efficiently cache a structured dataset

filtered result Observables now constructed on-demand from slices of structured internal cache

also abstracted out bindSocketClient() method to enable more flexible init and also potentially listening to updates from multiple Sensemaker Cells

pospi committed:
WIP finishing pass at revised data pipeline with merging of cached & new results in filtered streams

pospi commented:
@adaburrows I ended up following through with our conversation this morning RE the above. Have aligned the internal data model with an ES6 version of your proposal and, well, it *typechecks* so far. We will see what happens when I've done the other accessors but it's looking quite positive in b1b2426.

[L209](https://github.com/neighbour-hoods/sensemaker-lite/commit/b1b2426232ca05cbb15188dbb87cffbe0d0cf22d#diff-c81f75774fa1ca9e6ab11b0cabd31dd763a8b87d45d3d2c67fd02f56955ce42fR209) sets up the data pipeline for the internal cache; [L258](https://github.com/neighbour-hoods/sensemaker-lite/commit/b1b2426232ca05cbb15188dbb87cffbe0d0cf22d#diff-c81f75774fa1ca9e6ab11b0cabd31dd763a8b87d45d3d2c67fd02f56955ce42fR258) is *most* of what binding to filters over existing values looks like. I still have a little confusion to work through with merging the cached results with new ones; ba6fa16 being the attempt and a TODO which explains the remaining work. (I have some ideas, but any pointers are much appreciated.)

I've noticed ES6 `Map` and `Set` are probably sub-optimal unless transpilers do neat things, because in order to run useful processing you're generally doing `[...thing.values()]` to unpack data into an array *anyway* - so might change to ES5 `[]` & `{}` instead?
